File: 02/Exercise2/Program.cs
using System;
using System.IO;
using System.Linq;

namespace Exercise2
{
    class Program
    {
        static void Main(string[] args)
        {
            var lines = File.ReadAllLines(args[0]);
            var validPasswordCount = 0;
            foreach (var line in lines) {
                var lineParts = line.Split(" ");
                var positionParts = lineParts[0].Split("-");
                var firstPosition = int.Parse(positionParts[0]);
                var secondPosition = int.Parse(positionParts[1]);
                var character = lineParts[1][0];
                var password = lineParts[2];
                var characterCount = password.ToCharArray().Count(c => c == character);
                if (password[firstPosition-1] == character ^ password[secondPosition-1] == character) validPasswordCount++;
            }
            Console.WriteLine(validPasswordCount);
        }
    }
}

File: 02/fsharp/Exercise2/Program.fs
open System
open System.IO

type Policy = {
    ValidPositions: int * int;
    Letter: char;
}

let isValidPassword (policy, password: string) =
    let (firstPosition, secondPosition) = policy.ValidPositions
    (password.[firstPosition - 1] = policy.Letter) <> (password.[secondPosition - 1] = policy.Letter)

let parsePolicyAndPassword (text: string) =
    let parts = text.Split (" ")
    let [| firstPosition; secondPosition |] = parts.[0].Split ("-") |> Array.map int
    let letter = parts.[1] |> Seq.head
    let password = parts.[2]
    ({ ValidPositions = (firstPosition, secondPosition); Letter = letter }, password)

[<EntryPoint>]
let main argv =
    let input =
        File.ReadAllLines(argv.[0])
        |> Array.map parsePolicyAndPassword
    let validPasswords =
        input
        |> Array.filter isValidPassword
    Console.WriteLine(validPasswords.Length)
    0 // return an integer exit code
